pgml-extension/src/bindings/transformers/transformers.py

```diff
@@ -983,3 +997,617 @@ def generate(model_id, data, config):
             )
             all_preds.extend(decoded_preds)
     return all_preds
+
+
+#######################
+# LLM Fine-Tuning
+#######################
+
+
+class PGMLCallback(TrainerCallback):
+    "A callback that prints a message at the beginning of training"
+
+    def __init__(self, project_id, model_id):
+        self.project_id = project_id
+        self.model_id = model_id
+
+    def on_log(self, args, state, control, logs=None, **kwargs):
+        _ = logs.pop("total_flos", None)
+        if state.is_local_process_zero:
+            logs["step"] = state.global_step
+            logs["max_steps"] = state.max_steps
+            logs["timestamp"] = str(datetime.now())
+            print_info(json.dumps(logs, indent=4))
+            insert_logs(self.project_id, self.model_id, json.dumps(logs))
+
+
+class FineTuningBase:
+    def __init__(
+        self,
+        project_id: int,
+        model_id: int,
+        train_dataset: datasets.Dataset,
+        test_dataset: datasets.Dataset,
+        path: str,
+        hyperparameters: dict,
+    ) -> None:
+
+        # initialize class variables
+        self.project_id = project_id
+        self.model_id = model_id
+        self.train_dataset = train_dataset
+        self.test_dataset = test_dataset
+        self.token = None
+        self.load_in_8bit = False
+        self.tokenizer_args = None
+
+        # check if path is a directory
+        if not os.path.isdir(path):
+            os.makedirs(path, exist_ok=True)
+
+        self.path = path
+
+        # check if hyperparameters is a dictionary
+        if "model_name" not in hyperparameters:
+            raise ValueError("model_name is a required hyperparameter")
+        else:
+            self.model_name = hyperparameters.pop("model_name")
+
+        if "token" in hyperparameters:
```

Comment:
Isn't this a model init param, not a hyperparam, like many other things in this list? Maybe hyperparams covers everything?

Reply:
That's correct. Moved all the parameters to hyperparams.